src/chronify/sqlalchemy/functions.py

```diff
@@ -28,34 +47,51 @@ def write_database(
     config: TimeBaseModel,
     if_table_exists: DbWriteMode = "append",
 ) -> None:
-    """Write a Pandas DataFrame to the database.
-    For efficiency, this may mutate the input dataframe for datetime inputs. The expectation
-    is that the dataframe is already a copy of user inputs. This could be changed if needed.
-    """
-    if isinstance(config, DatetimeRange):
-        df = _convert_database_input_for_datetime(df, conn, config)
-    pl.DataFrame(df).write_database(table_name, connection=conn, if_table_exists=if_table_exists)
+    """Write a Pandas DataFrame to the database."""
+    match conn.engine.name:
+        case "duckdb":
+            assert conn._dbapi_connection is not None
+            assert conn._dbapi_connection.driver_connection is not None
+            match if_table_exists:
+                case "append":
+                    query = f"INSERT INTO {table_name} SELECT * FROM df"
+                case "replace":
+                    conn._dbapi_connection.driver_connection.sql(
+                        f"DROP TABLE IF EXISTS {table_name}"
+                    )
+                    query = f"CREATE TABLE {table_name} AS SELECT * FROM df"
+                case "fail":
+                    query = f"CREATE TABLE {table_name} AS SELECT * FROM df"
+                case _:
+                    msg = f"{if_table_exists=}"
+                    raise InvalidOperation(msg)
+            conn._dbapi_connection.driver_connection.sql(query)
+        case "sqlite":
+            if isinstance(config, DatetimeRange):
+                df = _convert_database_input_for_datetime(df, config)
+            pl.DataFrame(df).write_database(
+                table_name, connection=conn, if_table_exists=if_table_exists
+            )
+        case _:
+            pl.DataFrame(df).write_database(
+                table_name, connection=conn, if_table_exists=if_table_exists
+            )
 
 
-def _convert_database_input_for_datetime(
-    df: pd.DataFrame, conn: Connection, config: DatetimeRange
-) -> pd.DataFrame:
-    if (
-        conn.engine.name == "sqlite"
-        and isinstance(config, DatetimeRange)
-        and not config.is_time_zone_naive()
-    ):
+def _convert_database_input_for_datetime(df: pd.DataFrame, config: DatetimeRange) -> pd.DataFrame:
+    if config.is_time_zone_naive():
+        df2 = df
+    else:
+        df2 = df.copy()
```

Comment:
Any reason you need to make a copy here?

Reply:
With the recent additions, such as `Store.ingest_table`, the dataframe might be passed directly from the user. We shouldn't change their object.